sharedbuffers/mapped_struct.py

```diff
@@ -3005,22 +3005,55 @@ def register_subclass(cls, typ, supertyp):
         PROXY_TYPES[typ] = PROXY_TYPES[supertyp]
         return cls.OBJ_PACKERS[typecode][2]
 
+    @cython.locals(ivalue = cython.longlong)
     def __init__(self, value = None, typ = None):
         if value is None:
             self.typecode = None
             self.value = None
-        else:
-            if typ is None:
-                typ = type(value)
-            typ = TYPES.get(typ, typ)
-            if typ not in _mapped_object_TYPE_CODES and issubclass(typ, BufferProxyObject):
-                # Check bases
-                for base in typ.__bases__:
-                    if base is not object and base in _mapped_object_TYPE_CODES:
-                        typ = base
-                        break
-            self.typecode = _mapped_object_TYPE_CODES[typ]
-            self.value = value
+            return
+
+        if typ is None:
+            typ = type(value)
+
+        self.value = value
+
+        if typ is int or typ is long:
+            # Special case for numbers, to pick a smaller typecode when possible
+            try:
+                ivalue = value
```

Comment:
shouldn't this also set the typecode to 'q' ?

Reply:
That's done in the `else` part. This particular line is only trying to cast into a signed int64, but success here is not the criteria to actually assign typecode `q`.